### PR TITLE
Add an SCF/EFP reference implementation.

### PR DESCRIPTION
**Pull request by loriab**
## Description
Shows how to use libefp in an SCF/EFP sp

## What are your new additions? Please provide a brief list.
* **New Features**
  - [x] SCF/EFP sp

## Any questions for the community?
- [x] I'll have to rework the testing structure a bit and move to pytest to handle the very specific versions this requires cleanly. This will almost certainly break current testing infrastructure b/c needs recent psi4 and unreleased pylibefp.

## Status
- [x] Click when ready for review-and-merge. ~Review, sure. Merge, no.~

### CODE REVIEW

**Comment by loriab** [comment]
Ok, this has earned its green check with the new testing machinery. Ready for review.

File: Self-Consistent-Field/RHF_EFP.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+            atoms = efpobj.get_frag_atoms(ifr)
+            for iat in range(natoms[ifr]):
+                xyz_atom = [atoms[iat]['x'], atoms[iat]['y'], atoms[iat]['z']]
+                if np.allclose(xyz_atom, origin, atol=1e-10):
+                    val_mp[imp, 0] += atoms[iat]['Z']
+
+        # scale multipole integrals by multipole magnitudes. result goes into V
+        for pole in range(20):
+            efp_ints[pole] *= -prefacs[pole] * val_mp[imp, pole]
+            V2 += efp_ints[pole]
```
Comment (dgasmith) [inline Self-Consistent-Field/RHF_EFP.py]: Can always use broadcasting here. If I understand it correctly:

`V2 = np.sum(- efp_ints * val_mp[:, None], axis=0)`

File: Self-Consistent-Field/RHF_EFP.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+t = time.time()
+E = 0.0
+Enuc = mol.nuclear_repulsion_energy()
+Eold = 0.0
+Dold = np.zeros_like(D)
+
+for SCF_ITER in range(1, maxiter + 1):
+
+    # <-- efp: add contribution to Fock matrix
+    H = Horig
```
Comment (dgasmith) [inline Self-Consistent-Field/RHF_EFP.py]: Might be `H = Horig.copy()`.

File: Self-Consistent-Field/RHF_EFP.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    verbose_dipoles = 1 if (SCF_ITER == 1) else 0
+    # [Kaliman:2013:2284] Fig. 4 -- Compute electric field from wavefunction
+    # [Kaliman:2013:2284] Fig. 4 -- Compute electric field from induced dipoles
+    Vefp = modify_Fock_induced(nbf, efpmol, verbose=verbose_dipoles)
+    H = H + Vefp
```
Comment (dgasmith) [inline Self-Consistent-Field/RHF_EFP.py]: `H += Vefp`

File: Self-Consistent-Field/RHF_EFP.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+    diis_e = np.einsum('ij,jk,kl->il', F, D, S) - np.einsum('ij,jk,kl->il', S, D, F)
+    diis_e = A.dot(diis_e).dot(A)
+
+    # SCF energy and update
+    # [Kaliman:2013:2284] Fig. 4 -- Compute QM wavefunction
+    SCF_E = np.einsum('pq,pq->', F + H, D) + Enuc
+    dRMS = np.mean(diis_e**2)**0.5
+
+    # <-- efp: add contribution to energy
+    density = D
```
Comment (dgasmith) [inline Self-Consistent-Field/RHF_EFP.py]: What does this do here?

Comment (loriab) [reply Self-Consistent-Field/RHF_EFP.py]: It's a hack. The libefp callback requires a specific signature. Can't get the density into it through args, so it has to be named what the fn insides expect to access as "global". I'll explain it.

Comment (dgasmith) [reply Self-Consistent-Field/RHF_EFP.py]: Ah, I see. Maybe call it `efp_density` and inside that functional do a `global efp_density` to help make people aware?

Comment (loriab) [reply Self-Consistent-Field/RHF_EFP.py]: never had to use the dreaded `global` before. sure.

File: Self-Consistent-Field/RHF_EFP.py (hunk 1 above)
Comment (loriab) [reply Self-Consistent-Field/RHF_EFP.py]: I don't think so, as V2 is accumulated across both imp and pole loops, while efp_ints is redefined for each imp?

**Comment by loriab** [comment]
patched and rebased

**Pull request merged**